de.metas.shipper.gateway.dhl/src/main/java/de/metas/shipper/gateway/dhl/DhlShipperGatewayClient.java

```diff
@@ -0,0 +1,420 @@
+import de.metas.shipper.gateway.spi.model.DeliveryPosition;
+import de.metas.shipper.gateway.spi.model.PackageDimensions;
+import de.metas.shipper.gateway.spi.model.PackageLabel;
+import de.metas.shipper.gateway.spi.model.PackageLabels;
+import de.metas.shipper.gateway.spi.model.PickupDate;
+import lombok.Builder;
+import lombok.NonNull;
+import org.adempiere.exceptions.AdempiereException;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.oxm.jaxb.Jaxb2Marshaller;
+import org.springframework.ws.WebServiceMessage;
+import org.springframework.ws.client.core.WebServiceMessageCallback;
+import org.springframework.ws.client.core.WebServiceTemplate;
+import org.springframework.ws.soap.SoapHeader;
+import org.springframework.ws.soap.SoapMessage;
+import org.springframework.ws.transport.http.HttpComponentsMessageSender;
+import org.springframework.xml.transform.StringResult;
+
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBElement;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.time.format.DateTimeFormatter;
+import java.util.Base64;
+import java.util.List;
+
+public class DhlShipperGatewayClient implements ShipperGatewayClient
+{
+	private static final Logger logger = LoggerFactory.getLogger(DhlShipperGatewayClient.class);
+
+	private final Version API_VERSION;
+
+	private final DhlClientConfig config;
+
+	private final WebServiceTemplate webServiceTemplate;
+	private final SoapHeaderWithAuth soapHeaderWithAuth;
+
+	private final de.dhl.webservice.cisbase.ObjectFactory objectFactoryCis;
+	private final de.dhl.webservices.businesscustomershipping._3.ObjectFactory objectFactory;
+
+	@Builder
+	public DhlShipperGatewayClient(@NonNull final DhlClientConfig config)
+	{
+		this.config = config;
+
+		webServiceTemplate = createWebServiceTemplate();
+		soapHeaderWithAuth = new SoapHeaderWithAuth();
+
+		objectFactoryCis = new de.dhl.webservice.cisbase.ObjectFactory();
+		objectFactory = new de.dhl.webservices.businesscustomershipping._3.ObjectFactory();
+
+		// we're using DHL SOAP API V3
+		API_VERSION = objectFactory.createVersion();
+		API_VERSION.setMajorRelease("3");
+		API_VERSION.setMinorRelease("0");
+	}
+
+	@Override
+	public String getShipperGatewayId()
+	{
+		return DhlConstants.SHIPPER_GATEWAY_ID;
+	}
+
+	@Override
+	public DeliveryOrder createDeliveryOrder(final DeliveryOrder draftDeliveryOrder) throws ShipperGatewayException
+	{
+		throw new ShipperGatewayException("DRAFT Delivery Orders shall never be created.");
+	}
+
+	@Override
+	public DeliveryOrder completeDeliveryOrder(final DeliveryOrder deliveryOrder) throws ShipperGatewayException
+	{
+		logger.trace("Creating shipment order request for {}", deliveryOrder);
```

Comment:
hm, i think it might be a good idea to use `Loggable.withLogger()` thoughtout this client;
not neccesarily everywhere where you currently log..but at some places.
With loggable, the message will e.g. end up in the async-workpackage-log if the code is executed from there, or in AD_PInstance-log if the code is invoked from a process. This can be *very* helpful for our support-people